File: ruby/linked_list.rb
class Node
  def initialize(key, val)
    @key, @val = key, val
    @next, @prev = nil, nil
  end
  attr_accessor :key, :val, :next, :prev

  def inspect
    "(#{key}: #{val})"
  end
end

class LinkedList
  def initialize
    @head = Node.new(nil, nil)
    @tail = Node.new(nil, nil)

    @head.next = @tail
    @tail.prev = @head
  end

  def insert(key, val)
    node = Node.new(key, val)
    @tail.prev.next = node
    node.prev = @tail.prev
    node.next = @tail
    @tail.prev = node
  end

  def remove(key)
    node = find_node(key)
    return unless node
    node.prev.next = node.next
    node.next.prev = node.prev
    key
  end

  def find_node(key)
    self.find {|node| node.key == key}
  end

  include Enumerable
  def each(&prc)
    current = @head.next
    until current == @tail
      prc.call(current)
      current = current.next
    end
  end

  def inspect
    nodes = self.map {|node| "(#{node.key}: #{node.val})"}
    nodes.join(" -> ")
  end
end

File: javascript/linked_list.js
class Node {
  constructor(key, val) {
    this.key = key;
    this.val = val;
    this.prev = null;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = new Node(null, null);
    this.tail = new Node(null, null);

    this.head.next = this.tail;
    this.tail.prev = this.head;
  }

  get(key) {
    let current = this.head;

    while (current !== this.tail) {
      if (current.key === key) {return current.val;}
    }
  }

  insert(key, val) {
    let node = new Node(key, val);

    this.tail.prev.next = node;
    node.prev = this.tail.prev;
    this.tail.prev = node;
    node.next = this.tail;
    return key;
  }

  delete(key) {
    let current = this.head;

    while (current !== this.tail) {
      if (current.key === key) {
        current.prev.next = current.next;
        current.next.prev = current.prev;
        break;
      }
      current = current.next;
    }
    return current.key;
  }

  contains(key) {
    let current = this.head.next;
    while (current !== this.tail ) {
      if (current.key === key) {return true;}
      current = current.next;
    }
    return false;
  }

  draw() {
    let list = [];
    this.each( node => list.push([node.key, node.val]));
    return list.join(" -> ");
  }

  each(callback) {
    var current = this.head.next;
    while (current !== this.tail) {
      callback(current);
      current = current.next;
    }
  }
}
